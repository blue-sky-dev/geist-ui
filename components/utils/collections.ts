import React, { ReactNode } from 'react'

export const getId = () => {
  return Math.random().toString(32).slice(2, 10)
}

export const hasChild = (
  children: ReactNode | undefined,
  child: React.ElementType
): boolean => {
  const types = React.Children.map(children, item => {
    if (!React.isValidElement(item)) return null
    return item.type
  })
  
  return (types || []).includes(child)
}

export const pickChild = (
  children: ReactNode | undefined,
  targetChild: React.ElementType
): [ReactNode | undefined, ReactNode | undefined] => {
  let target: ReactNode[] = []
  const withoutTargetChildren = React.Children.map(children, item => {
    if (!React.isValidElement(item)) return item
    if (item.type === targetChild) {
      target.push(item)
      return null
    }
    return item
  })
  
  const targetChildren = target.length >= 0 ? target : undefined
  
  return [withoutTargetChildren, targetChildren]
}

export const pickChildByProps = (
  children: ReactNode | undefined,
  key: string,
  value: any,
): [ReactNode | undefined, ReactNode | undefined] => {
  let target: ReactNode[] = []
  const withoutPropChildren = React.Children.map(children, item => {
    if (!React.isValidElement(item)) return null
    if (!item.props) return item
    if (item.props[key] === value) {
      target.push(item)
      return null
    }
    return item
  })
  
  const targetChildren = target.length >= 0 ? target : undefined
  
  return [withoutPropChildren, targetChildren]
}

export const pickChildrenFirst = (
  children: ReactNode | undefined,
): ReactNode | undefined => {
  return React.Children.toArray(children)[0]
}

export const setChildrenProps = (
  children: ReactNode | undefined,
  props: object = {},
  targetComponents: Array<React.ElementType> = [],
): ReactNode | undefined => {
  if (React.Children.count(children) === 0) return []
  const allowAll = targetComponents.length === 0
  const clone = (child: React.ReactElement, props = {}) => React.cloneElement(child, props)

  return React.Children.map(children, item => {
    if (!React.isValidElement(item)) return item
    if (allowAll) return clone(item, props)

    const isAllowed = targetComponents.find(child => child === item.type)
    if (isAllowed) return clone(item, props)
    return item
  })
}

export const setChildrenIndex = (
  children: ReactNode | undefined,
  targetComponents: Array<React.ElementType> = [],
): ReactNode | undefined => {
  if (React.Children.count(children) === 0) return []
  const allowAll = targetComponents.length === 0
  const clone = (child: React.ReactElement, props = {}) => React.cloneElement(child, props)
  let index = 0
  
  return React.Children.map(children, item => {
    if (!React.isValidElement(item)) return item
    index = index + 1
    if (allowAll) return clone(item, { index })
    
    const isAllowed = targetComponents.find(child => child === item.type)
    if (isAllowed) return clone(item, { index })
    index = index - 1
    return item
  })
}

export type ShapeType = {
  width: number
  height: number
}

export const getRealShape = (el: HTMLElement | null): ShapeType => {
  const defaultShape: ShapeType = { width: 0, height: 0 }
  if (!el || typeof window === 'undefined') return defaultShape
  
  const rect = el.getBoundingClientRect()
  const { width, height } = window.getComputedStyle(el)
  
  const getCSSStyleVal = (str: string, parentNum: number) => {
    if (!str) return 0
    const strVal = str.includes('px') ? +str.split('px')[0]
      : str.includes('%') ? +str.split('%')[0] * parentNum * 0.01 : str
    
    return Number.isNaN(+strVal) ? 0 : +strVal
  }

  return {
    width: getCSSStyleVal(`${width}`, rect.width),
    height: getCSSStyleVal(`${height}`, rect.height),
  }
}
