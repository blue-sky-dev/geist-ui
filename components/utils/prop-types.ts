
export const tuple = <T extends string[]>(...args: T) => args

const buttonTypes = tuple(
  'default',
  'secondary',
  'success',
  'warning',
  'error',
  'abort',
)

const normalSizes = tuple(
  'mini',
  'small',
  'medium',
  'large',
)

const normalTypes = tuple(
  'default',
  'secondary',
  'success',
  'warning',
  'error',
)

const themeTypes = tuple(
  'dark',
  'light',
)

const snippetTypes = tuple(
  'default',
  'secondary',
  'success',
  'warning',
  'error',
  'dark',
  'lite',
)

const copyTypes = tuple(
  'default',
  'slient',
  'prevent',
)

export type ButtonTypes = typeof buttonTypes[number]

export type NormalSizes = typeof normalSizes[number]

export type NormalTypes = typeof normalTypes[number]

export type ThemeTypes = typeof themeTypes[number]

export type SnippetTypes = typeof snippetTypes[number]

export type CopyTypes = typeof copyTypes[number]
