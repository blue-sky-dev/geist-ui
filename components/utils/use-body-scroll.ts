import useBodyScroll from '../utils-shared-deprecated/use-body-scroll'

export default useBodyScroll
