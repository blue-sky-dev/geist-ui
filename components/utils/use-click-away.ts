import { default as useClickAway } from '../utils-shared/use-click-away'

export default useClickAway
