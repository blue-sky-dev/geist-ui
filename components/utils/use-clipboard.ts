import useClipboard from '../utils-shared/use-clipboard'

export default useClipboard
