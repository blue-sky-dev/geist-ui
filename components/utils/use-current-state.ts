import useCurrentState from '../utils-shared/use-current-state'

export default useCurrentState
