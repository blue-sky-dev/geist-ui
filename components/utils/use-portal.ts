import { useEffect, useState } from 'react'
import useSSR from '../utils/use-ssr'

const getId = () => {
  return Math.random().toString(32).slice(2, 10)
}

const createElement = (id: string): HTMLElement => {
  const el = document.createElement('div')
  el.setAttribute('id', id)
  return el
}

const usePortal = (selectId: string = getId()): Element | null => {
  const id = `zeit-ui-${selectId}`
  const { isBrowser } = useSSR()
  const [elSnapshot, setElSnapshot] = useState<Element | null>(isBrowser ? createElement(id) : null)
  
  useEffect(() => {
    const hasElement = document.querySelector(`#${id}`)
    const el = hasElement || createElement(id)

    if (!hasElement) {
      document.body.appendChild(el)
    }
    setElSnapshot(el)
  
    return () => {
      const node = document.getElementById(id)
      if (node) {
        // document.body.removeChild(node)
      }
    }
  }, [])

  return elSnapshot
}

export default usePortal
