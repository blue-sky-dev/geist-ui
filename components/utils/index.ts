import { default as useBodyScroll } from './use-body-scroll'
import { default as useClickAnywhere } from './use-click-anywhere'
import { default as useClickAway } from './use-click-away'
import { default as useClipboard } from './use-clipboard'
import { default as useCurrentState } from './use-current-state'
import { default as useRealShape } from './use-real-shape'
import { default as useResize } from './use-resize'
import { default as useSSR } from './use-ssr'
import { default as withDefaults } from './with-defaults'

export default {
  withDefaults,
  useBodyScroll,
  useClickAnywhere,
  useClickAway,
  useClipboard,
  useCurrentState,
  useRealShape,
  useResize,
  useSSR,
}
