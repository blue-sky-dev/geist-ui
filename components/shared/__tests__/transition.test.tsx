import React from 'react'
import { mount } from 'enzyme'
import CSSTransition from '../css-transition'
import { updateWrapper } from 'tests/utils'

describe('CSSTransition', () => {
  it('should render correctly', () => {
    const wrapper = mount(
      <CSSTransition visible>
        <span>test</span>
      </CSSTransition>,
    )
    expect(wrapper.text()).toContain('test')
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => wrapper.unmount()).not.toThrow()
  })

  it('should work correctly with time props', async () => {
    const wrapper = mount(
      <CSSTransition enterTime={300} leaveTime={300}>
        <span id="test">test</span>
      </CSSTransition>,
    )
    expect(wrapper.find('.transition-enter-active').length).toBe(0)

    wrapper.setProps({ visible: true })
    await updateWrapper(wrapper, 310)
    expect(wrapper.find('.transition-enter-active').length).not.toBe(0)

    wrapper.setProps({ visible: false })
    await updateWrapper(wrapper, 310)
    expect(wrapper.find('.transition-leave-active').length).not.toBe(0)
  })

  it('should clear css-transition classes after hidden', async () => {
    const wrapper = mount(
      <CSSTransition visible>
        <span>test</span>
      </CSSTransition>,
    )
    // don't remove classes after shown
    await updateWrapper(wrapper, 60)
    expect(wrapper.find('.transition-enter-active').length).not.toBe(0)

    await updateWrapper(wrapper, 150)
    expect(wrapper.find('.transition-enter-active').length).not.toBe(0)

    // remove classes after hidden
    wrapper.setProps({ visible: false })
    await updateWrapper(wrapper, 60)
    expect(wrapper.find('.transition-leave-active').length).not.toBe(0)

    await updateWrapper(wrapper, 150)
    expect(wrapper.find('.transition-leave-active').length).toBe(0)
    expect(wrapper.find('.transition-enter-active').length).toBe(0)
  })

  it('custom class names should be rendered', async () => {
    const wrapper = mount(
      <CSSTransition name="test">
        <span id="test">test</span>
      </CSSTransition>,
    )

    expect(wrapper.find('.test-enter-active').length).toBe(0)

    wrapper.setProps({ visible: true })
    await updateWrapper(wrapper, 60)
    expect(wrapper.find('.test-enter-active').length).not.toBe(0)

    wrapper.setProps({ visible: false })
    await updateWrapper(wrapper, 60)
    expect(wrapper.find('.test-leave-active').length).not.toBe(0)
  })
})
