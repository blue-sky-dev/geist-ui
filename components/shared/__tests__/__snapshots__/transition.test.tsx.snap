// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSSTransition should render correctly 1`] = `"<span class=\\"undefined  transition-enter\\">test</span>"`;
